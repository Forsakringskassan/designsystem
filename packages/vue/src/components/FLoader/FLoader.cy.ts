import { defineComponent } from "vue";
import { FLoaderPageObject } from "../../cypress";
import FLoader from "./FLoader.vue";

const loader = new FLoaderPageObject();

describe("FLoader", () => {
    it("should have a page object that can access any necessary elements", () => {
        cy.mount(FLoader, { props: { show: true } });
        loader.el().should("be.visible");
        loader.waitText().should("have.trimmedText", "Vänligen vänta");
    });

    it("should use custom text if provided", () => {
        cy.mount(FLoader, {
            props: {
                show: true,
            },
            slots: {
                default: "Lorem ipsum dolor sit amet",
            },
        });
        loader.el().should("be.visible");
        loader.waitText().should("contain.text", "Lorem ipsum dolor sit amet");
    });

    it("should be displayed after a 1s when delay is enabled", () => {
        cy.mount(FLoader, {
            props: {
                show: true,
                overlay: true,
                delay: true,
            },
        });
        // eslint-disable-next-line cypress/no-unnecessary-waiting -- needed to acually wait. cy.clock() does not work with css-animations.
        cy.wait(500);
        loader.wrapper().should("not.be.visible");
        // eslint-disable-next-line cypress/no-unnecessary-waiting -- needed to acually wait. cy.clock() does not work with css-animations.
        cy.wait(1000);
        loader.wrapper().should("be.visible");
        loader.waitText().should("have.focus");
    });

    describe("Overlay Mode", () => {
        it("should have focus on loading text when opened", () => {
            cy.mount(FLoader, {
                props: {
                    show: true,
                    overlay: true,
                },
            });
            loader.wrapper().should("be.visible");
            loader.waitText().should("have.focus");
        });

        it("should restore focus after tooltip is closed", () => {
            const TestComponent = defineComponent({
                components: { FLoader },
                template: /* HTML */ `
                    <button type="button" @click="onClick">enable</button>
                    <f-loader :show overlay></f-loader>
                `,
                data() {
                    return {
                        show: false,
                    };
                },
                methods: {
                    onClick() {
                        this.show = true;
                        setTimeout(() => {
                            this.show = false;
                        }, 2000);
                    },
                },
            });
            cy.mount(TestComponent);
            cy.get("button").click();
            loader.wrapper().should("be.visible");
            loader.waitText().should("have.focus");
            cy.get("button").should("have.focus");
        });
    });

    describe("teleport", () => {
        // A teleported loader should be a direct child of element given in `config.teleportTarget`.
        const teleportedLoader = new FLoaderPageObject("#teleport > .loader");

        it("should use teleport if using `overlay`", () => {
            cy.mount(FLoader, {
                props: {
                    show: true,
                    overlay: true,
                },
            });
            teleportedLoader.el().should("exist");
        });

        it("should not use teleport if not using `overlay`", () => {
            cy.mount(FLoader, {
                props: {
                    show: true,
                },
            });
            teleportedLoader.el().should("not.exist");
        });
    });
});
