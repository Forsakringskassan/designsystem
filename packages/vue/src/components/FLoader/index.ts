export { default as FLoader } from "./FLoader.vue";
