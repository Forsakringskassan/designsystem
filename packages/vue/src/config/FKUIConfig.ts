/**
 * @public
 */
export enum FKUIConfigButtonOrder {
    LEFT_TO_RIGHT,
    RIGHT_TO_LEFT,
}

/**
 * @public
 */
export interface FKUIConfig {
    buttonOrder: FKUIConfigButtonOrder;
    teleportTarget: string | Element;
    /**
     * @deprecated Use `teleportTarget` instead.
     */
    modalTarget: string | Element | null;
    /**
     * @deprecated Use `teleportTarget` instead.
     */
    popupTarget: string | Element | null;
    get popupContainer(): HTMLElement;
    set popupContainer(value: string | Element);
    production: boolean;
}
