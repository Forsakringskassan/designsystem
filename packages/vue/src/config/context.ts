import { type App, type AppContext } from "vue";

declare module "@vue/runtime-core" {
    interface ComponentCustomProperties {
        $fkui?: FKUIContext;
    }
}

/**
 * @public
 */
export interface FKUIContext {
    appContext: AppContext;
}

/**
 * @public
 */
export interface MaybeWithFKUIContext {
    $fkui?: FKUIContext;
}

/**
 * @public
 * @param app - Running app
 */
export function setRunningContext(app: App): void {
    const fkuiContext: FKUIContext = {
        appContext: app._context,
    };

    app.config.globalProperties.$fkui = fkuiContext;
}

/**
 * @internal
 * @param callingInstance - Calling component
 */
export function getRunningContext(
    callingInstance: MaybeWithFKUIContext,
): FKUIContext {
    if (!callingInstance.$fkui) {
        throw new Error(
            "Application running context is unset. Call `setRunningContext(app)` after `app = createApp(..)`.",
        );
    }

    return callingInstance.$fkui;
}
