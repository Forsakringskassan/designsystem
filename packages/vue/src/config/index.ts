export { config } from "./config";
export { type FKUIConfig, FKUIConfigButtonOrder } from "./FKUIConfig";
export {
    setRunningContext,
    getRunningContext,
    type FKUIContext,
    type MaybeWithFKUIContext,
} from "./context";
