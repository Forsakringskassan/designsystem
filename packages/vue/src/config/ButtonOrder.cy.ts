import FCrudDatasetTableExample from "../components/FCrudDataset/examples/FCrudDatasetTableExample.vue";
import FConfirmModalApiExample from "../components/FModal/examples/FConfirmModalApiExample.vue";
import FFormModalApiExample from "../components/FModal/examples/FFormModalApiExample.vue";
import { FCrudDatasetPageObject, FModalPageObject } from "../cypress";
import { config } from "./config";
import { FKUIConfigButtonOrder } from "./FKUIConfig";

describe("Components", () => {
    describe("FConfirmModal", () => {
        describe("LEFT_TO_RIGHT", () => {
            it("should respect button order", () => {
                config.buttonOrder = FKUIConfigButtonOrder.LEFT_TO_RIGHT;
                cy.mount(FConfirmModalApiExample);
                cy.get("button").click();
                cy.get(".button-group").should(
                    "contain.text",
                    "Ja, ta bort Nej, behåll",
                );
                const modal = new FModalPageObject(".modal");
                modal.closeCross().click();
            });
        });

        describe("RIGHT_TO_LEFT", () => {
            it("should respect button order", () => {
                config.buttonOrder = FKUIConfigButtonOrder.RIGHT_TO_LEFT;
                cy.mount(FConfirmModalApiExample);
                cy.get("button").click();
                cy.get(".button-group").should(
                    "contain.text",
                    "Nej, behåll Ja, ta bort",
                );
                const modal = new FModalPageObject(".modal");
                modal.closeCross().click();
            });
        });
    });

    describe("FFormModal", () => {
        describe("LEFT_TO_RIGHT", () => {
            it("should always show buttons left to right", () => {
                config.buttonOrder = FKUIConfigButtonOrder.LEFT_TO_RIGHT;
                cy.mount(FFormModalApiExample);
                cy.get("button").click();
                cy.get(".button-group").should("contain.text", "Spara Avbryt");
                const modal = new FModalPageObject(".modal");
                modal.closeCross().click();
            });
        });

        describe("RIGHT_TO_LEFT", () => {
            it("should ignore button order", () => {
                config.buttonOrder = FKUIConfigButtonOrder.RIGHT_TO_LEFT;
                cy.mount(FFormModalApiExample);
                cy.get("button").click();
                cy.get(".button-group").should("contain.text", "Spara Avbryt");
                const modal = new FModalPageObject(".modal");
                modal.closeCross().click();
            });
        });
    });

    describe("FCrudDataset", () => {
        describe("Lägg till", () => {
            describe("LEFT_TO_RIGHT", () => {
                it("should always show buttons left to right", () => {
                    config.buttonOrder = FKUIConfigButtonOrder.LEFT_TO_RIGHT;
                    cy.mount(FCrudDatasetTableExample);
                    const crudDataset = new FCrudDatasetPageObject(
                        ".crud-dataset",
                    );
                    crudDataset.addButton().click();
                    const modal = new FModalPageObject(".modal");
                    modal
                        .el()
                        .get(".button-group")
                        .should("contain.text", "Lägg till Avbryt");
                    modal.closeCross().click();
                });
            });

            describe("RIGHT_TO_LEFT", () => {
                it("should ignore button order", () => {
                    config.buttonOrder = FKUIConfigButtonOrder.RIGHT_TO_LEFT;
                    cy.mount(FCrudDatasetTableExample);
                    const crudDataset = new FCrudDatasetPageObject(
                        ".crud-dataset",
                    );
                    crudDataset.addButton().click();
                    const modal = new FModalPageObject(".modal");
                    modal
                        .el()
                        .get(".button-group")
                        .should("contain.text", "Lägg till Avbryt");
                    modal.closeCross().click();
                });
            });
        });

        describe("Ändra", () => {
            describe("LEFT_TO_RIGHT", () => {
                it("should always show buttons left to right", () => {
                    config.buttonOrder = FKUIConfigButtonOrder.LEFT_TO_RIGHT;
                    cy.mount(FCrudDatasetTableExample);
                    cy.get(".f-icon-pen:nth(0)").click({ force: true });
                    const modal = new FModalPageObject(".modal");
                    modal
                        .el()
                        .get(".button-group")
                        .should("contain.text", "Spara Avbryt");
                    modal.closeCross().click();
                });
            });

            describe("RIGHT_TO_LEFT", () => {
                it("should ignore button order", () => {
                    config.buttonOrder = FKUIConfigButtonOrder.RIGHT_TO_LEFT;
                    cy.mount(FCrudDatasetTableExample);
                    cy.get(".f-icon-pen:nth(0)").click({ force: true });
                    const modal = new FModalPageObject(".modal");
                    modal
                        .el()
                        .get(".button-group")
                        .should("contain.text", "Spara Avbryt");
                    modal.closeCross().click();
                });
            });
        });

        describe("Ta bort", () => {
            describe("LEFT_TO_RIGHT", () => {
                it("should respect button order", () => {
                    config.buttonOrder = FKUIConfigButtonOrder.LEFT_TO_RIGHT;
                    cy.mount(FCrudDatasetTableExample);
                    cy.get(".f-icon-trashcan:nth(0)").click({ force: true });
                    const modal = new FModalPageObject(".modal");
                    modal
                        .el()
                        .get(".button-group")
                        .should("contain.text", "Ja, ta bort Nej, avbryt");
                    modal.closeCross().click();
                });
            });

            describe("RIGHT_TO_LEFT", () => {
                it("should respect button order", () => {
                    config.buttonOrder = FKUIConfigButtonOrder.RIGHT_TO_LEFT;
                    cy.mount(FCrudDatasetTableExample);
                    cy.get(".f-icon-trashcan:nth(0)").click({ force: true });
                    const modal = new FModalPageObject(".modal");
                    modal
                        .el()
                        .get(".button-group")
                        .should("contain.text", "Nej, avbryt Ja, ta bort");
                    modal.closeCross().click();
                });
            });
        });
    });
});
