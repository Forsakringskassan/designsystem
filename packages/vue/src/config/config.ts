import { configLogic } from "@fkui/logic";
import { type FKUIConfig, FKUIConfigButtonOrder } from "./FKUIConfig";

let popupContainer: string | HTMLElement = document.body;
let production = true;

/**
 * @public
 */
export const config: FKUIConfig = {
    buttonOrder: FKUIConfigButtonOrder.RIGHT_TO_LEFT,
    teleportTarget: document.body,

    get popupContainer(): HTMLElement {
        if (typeof popupContainer === "string") {
            const element = document.querySelector<HTMLElement>(popupContainer);
            if (!element) {
                throw new Error(
                    `Failed to find popupContainer element from selector "${popupContainer}"`,
                );
            }
            return element;
        } else {
            return popupContainer;
        }
    },

    set popupContainer(value: string | HTMLElement) {
        popupContainer = value;
    },

    set production(value: boolean) {
        production = value;
        configLogic.production = value;
    },
    get production(): boolean {
        return production;
    },
};
