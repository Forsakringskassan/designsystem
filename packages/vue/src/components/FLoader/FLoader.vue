<script lang="ts">
import { type PropType, defineComponent } from "vue";
import { addFocusListener, findTabbableElements, removeFocusListener, restoreFocus, saveFocus } from "@fkui/logic";
import { TranslationMixin } from "../../plugins";
import { config } from "../../config";
import { focus } from "../../utils";

export default defineComponent({
    name: "FLoader",
    mixins: [TranslationMixin],
    inheritAttrs: false,
    props: {
        /* Aria-live must always be visible, hence v-if on component level is not possible, therefore we use show-prop */
        /**
         * Determine if the loader is visible or not
         */
        show: {
            type: Boolean,
            required: true,
            default: false,
        },
        /**
         * If loader should be displayed as a fullscreen overlay.
         */
        overlay: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * Delay the loader icon and text by 1 second
         */
        delay: {
            type: Boolean,
            required: false,
            default: false,
        },
        /**
         * Language used for determining fallback value for the loading text. Useful if loader is displayed before
         * text keys have been downloaded
         */
        language: {
            type: String,
            required: false,
            default: "sv",
        },
        /**
         * Set teleport target (when overlay is enabled).
         *
         * - When set to a string it is used as a selector.
         * - When set to a element it is used directly.
         * - Default uses {@link config#teleportTarget}.
         */
        teleport: {
            type: [String, HTMLElement] as PropType<string | HTMLElement | undefined>,
            required: false,
            default: undefined,
        },
    },
    data() {
        return {
            oldFocus: undefined as unknown as HTMLElement,
        };
    },
    computed: {
        defaultLoadingText(): string {
            return this.language === "en" ? "Please wait" : "Vänligen vänta";
        },
        classes(): Record<string, boolean> {
            return {
                "loader--overlay": this.overlay,
                "loader--delay": this.delay,
            };
        },
        teleportTarget() {
            if (this.teleport) {
                return this.teleport;
            } else {
                return config.teleportTarget;
            }
        },
        teleportDisabled() {
            return !this.overlay;
        },
    },
    watch: {
        show(show) {
            if (show) {
                this.openLoader();
            } else {
                this.closeLoader();
            }
        },
    },
    mounted() {
        if (this.show) {
            this.openLoader();
        }
    },
    methods: {
        async listener(): Promise<void> {
            await this.$nextTick();
            focus(this.$refs["loader-text"]);
        },
        async openLoader() {
            if (this.overlay) {
                saveFocus(document);
                this.listener();
                addFocusListener(findTabbableElements(document), this.listener);
            }
        },
        closeLoader() {
            if (this.overlay) {
                removeFocusListener(findTabbableElements(document), this.listener);
                restoreFocus();
            }
        },
    },
});
</script>

<template>
    <teleport :to="teleportTarget" :disabled="teleportDisabled">
        <div v-show="show" v-bind="$attrs" class="loader" :class="classes">
            <div class="loader__backdrop">
                <div class="loader__wrapper">
                    <div class="loader__spinner-1 loader__spinner">
                        <div class="loader__spinner-1-circle1 loader__circle loader__circle--1"></div>
                        <div class="loader__spinner-1-circle2 loader__circle loader__circle--2"></div>
                        <div class="loader__spinner-1-circle3 loader__circle loader__circle--3"></div>
                        <div class="loader__spinner-1-circle4 loader__circle loader__circle--4"></div>
                    </div>
                    <div class="loader__spinner-2 loader__spinner">
                        <div class="loader__spinner-2-circle1 loader__circle loader__circle--1"></div>
                        <div class="loader__spinner-2-circle2 loader__circle loader__circle--2"></div>
                        <div class="loader__spinner-2-circle3 loader__circle loader__circle--3"></div>
                        <div class="loader__spinner-2-circle4 loader__circle loader__circle--4"></div>
                    </div>
                    <div class="loader__spinner-3 loader__spinner">
                        <div class="loader__spinner-3-circle1 loader__circle loader__circle--1"></div>
                        <div class="loader__spinner-3-circle2 loader__circle loader__circle--2"></div>
                        <div class="loader__spinner-3-circle3 loader__circle loader__circle--3"></div>
                        <div class="loader__spinner-3-circle4 loader__circle loader__circle--4"></div>
                    </div>
                </div>
                <div ref="loader-text" class="loader__wait-text" :class="{ 'loader--delay': delay }" tabindex="-1">
                    <span role="alert">
                        <!-- @slot Slot for define a custom loading text -->
                        <slot>{{ $t("fkui.loader.wait.text", defaultLoadingText) }}</slot>
                    </span>
                </div>
            </div>
        </div>
    </teleport>
</template>
