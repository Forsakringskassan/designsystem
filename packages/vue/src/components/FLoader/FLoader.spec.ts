import "html-validate/jest";
import { shallowMount } from "@vue/test-utils";
import { config } from "../../config";
import FLoader from "./FLoader.vue";

describe("FLoader", () => {
    it("should not have overlay by default", () => {
        const wrapper = shallowMount(FLoader, {
            props: { overlay: false, show: false },
            global: {
                stubs: ["teleport"],
            },
        });
        expect(wrapper.find(".loader--overlay").exists()).toBeFalsy();
    });

    it("should show overlay", () => {
        const wrapper = shallowMount(FLoader, {
            props: { overlay: true, show: false },
            global: {
                stubs: ["teleport"],
            },
        });
        expect(wrapper.get(".loader--overlay")).toBeTruthy();
    });

    it("should have text by default", () => {
        const wrapper = shallowMount(FLoader, {
            props: { overlay: true, show: false },
            global: {
                stubs: ["teleport"],
            },
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("should be able to specify text", () => {
        const wrapper = shallowMount(FLoader, {
            props: { show: false },
            slots: { default: `Loading...` },
            global: {
                stubs: ["teleport"],
            },
        });
        expect(wrapper.get(".loader__wait-text")).toBeTruthy();
    });

    it("should show the specified text", () => {
        const wrapper = shallowMount(FLoader, {
            props: { show: true },
            slots: { default: `Loading...` },
            global: {
                stubs: ["teleport"],
            },
        });
        expect(wrapper.get(".loader__wait-text")).toBeTruthy();
    });

    it("should show loading text in english", () => {
        const wrapper = shallowMount(FLoader, {
            props: { show: true, language: "en" },
            global: {
                stubs: ["teleport"],
            },
        });
        expect(wrapper.get(".loader__wait-text").text()).toBe("Please wait");
    });
});

describe("props", () => {
    describe("teleport", () => {
        it("should teleport config.teleportTarget by default", () => {
            expect.assertions(1);
            config.teleportTarget = "#selector";
            const wrapper = shallowMount(FLoader, {
                props: { show: true },
                global: {
                    stubs: {
                        teleport: {
                            props: ["to", "disabled"],
                            template: "{{ to }}",
                        },
                    },
                },
            });
            expect(wrapper.text()).toBe("#selector");
        });

        it("should teleport to target set by prop", () => {
            expect.assertions(1);
            config.teleportTarget = "#selector";
            const wrapper = shallowMount(FLoader, {
                props: { show: true, teleport: "#overriden" },
                global: {
                    stubs: {
                        teleport: {
                            props: ["to", "disabled"],
                            template: "{{ to }}",
                        },
                    },
                },
            });
            expect(wrapper.text()).toBe("#overriden");
        });
    });
});

describe("html-validate", () => {
    it.each`
        html
        ${"<f-loader></f-loader>"}
        ${"<f-loader>Loader</f-loader>"}
        ${"<f-loader><template> Read more about FLoader</template></f-loader>"}
    `("$html should be valid", ({ html }) => {
        expect.assertions(1);

        expect(html).toHTMLValidate();
    });

    it.each`
        html
        ${"<f-loader><div></div></f-loader>"}
        ${"<f-loader><p></p></f-loader>"}
        ${"<f-loader><button></button></f-loader>"}
    `("$html should be invalid", ({ html }) => {
        expect.assertions(3);
        let catchedError;

        try {
            expect(html).toHTMLValidate();
        } catch (error) {
            catchedError = error;
        } finally {
            expect(catchedError).toBeDefined();
            expect(catchedError).toMatchSnapshot();
        }
    });
});
