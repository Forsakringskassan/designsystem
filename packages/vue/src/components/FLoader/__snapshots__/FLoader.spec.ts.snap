// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FLoader should have text by default 1`] = `
<teleport-stub to="[object HTMLBodyElement]" disabled="false">
  <div class="loader loader--overlay" style="display: none;">
    <div class="loader__backdrop">
      <div class="loader__wrapper">
        <div class="loader__spinner-1 loader__spinner">
          <div class="loader__spinner-1-circle1 loader__circle loader__circle--1"></div>
          <div class="loader__spinner-1-circle2 loader__circle loader__circle--2"></div>
          <div class="loader__spinner-1-circle3 loader__circle loader__circle--3"></div>
          <div class="loader__spinner-1-circle4 loader__circle loader__circle--4"></div>
        </div>
        <div class="loader__spinner-2 loader__spinner">
          <div class="loader__spinner-2-circle1 loader__circle loader__circle--1"></div>
          <div class="loader__spinner-2-circle2 loader__circle loader__circle--2"></div>
          <div class="loader__spinner-2-circle3 loader__circle loader__circle--3"></div>
          <div class="loader__spinner-2-circle4 loader__circle loader__circle--4"></div>
        </div>
        <div class="loader__spinner-3 loader__spinner">
          <div class="loader__spinner-3-circle1 loader__circle loader__circle--1"></div>
          <div class="loader__spinner-3-circle2 loader__circle loader__circle--2"></div>
          <div class="loader__spinner-3-circle3 loader__circle loader__circle--3"></div>
          <div class="loader__spinner-3-circle4 loader__circle loader__circle--4"></div>
        </div>
      </div>
      <div class="loader__wait-text" tabindex="-1"><span role="alert">Vänligen vänta</span></div>
    </div>
  </div>
</teleport-stub>
`;

exports[`html-validate <f-loader><button></button></f-loader> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <button> is missing recommended "type" attribute [no-implicit-button-type]
  <button> must have accessible text [text-content]]
`;

exports[`html-validate <f-loader><div></div></f-loader> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <div> element is not permitted as content under <f-loader> [element-permitted-content]]
`;

exports[`html-validate <f-loader><p></p></f-loader> should be invalid 1`] = `
[Error: Expected HTML to be valid but had the following errors:

  <p> element is not permitted as content under <f-loader> [element-permitted-content]]
`;
